py/server/deephaven/pandas.py

```diff
@@ -141,6 +182,20 @@ def to_table(df: pandas.DataFrame, cols: List[str] = None) -> Table:
             if diff_set:
                 raise DHError(message=f"columns - {list(diff_set)} not found")
 
+        # check if all the cols are pyarrow-backed and if so create an intermediate pyarrow table to upload to DH
+        df_dtypes = [df[col].dtype for col in cols]
+        if all(_is_pyarrow_backed(dt) for dt in df_dtypes):
+            pa_arrays = []
+            for col in cols:
+                pa_arrays.append(df[col].array)
+            pa_table = pa.Table.from_arrays(arrays=pa_arrays, names=cols)
+            dh_table = arrow.to_table(pa_table)
+            del pa_table
+            return dh_table
+
+        if any(_is_pyarrow_backed(dt) for dt in df_dtypes):
+            raise DHError(message="to_table doesn't support a mixture of pyarrow-backed arrays and numpy arrays")
```

Comment:
Again pa.Table.from_pandas() can cover some cases but due to the reasons listed above, it can become quite messy when type mapping and null handling is involved. Also it is not clear that the situation checked really exists in the real world because 1) the pyarrow dtype_backend is only available in ingesting data into Pandas, 2) it does the applications no favor to have mixed backend arrays as far as consistency and cleanness is concerned.

Reply:
Do you get to this state if you add a new column to a pyarrow df?